### ISSUE TITLE
Add validated delete for Roles

### ISSUE THREAD

**Issue by nivemaham** [RADAR-CNS/ManagementPortal#13]
Currently a user can be created based on a role. A role is a unique combination of Project and Authority. 
If a Role has to be removed, a prior validations should be made to check whether any User entities are created for specific role. If not allow detete-role for super-super.

**Comment by blootsvoets** [comment]
Why would you allow to delete roles? I thought that within a certain release of the management portals, the roles would be fixed?

**Comment by nivemaham** [comment]
Indeed. The only reason we would want to allow  removing a role is in case if there is any role, that was created by mistake. To support "Human errors". Currently. A role cannot be deleted.

**Comment by blootsvoets** [comment]
Ok I misunderstood how roles are implemented. What I would expect is that there is a direct link

- user `A1` has authority `SYS_ADMIN` on management portal (`null`?)
- user `A1` has authority `PROJECT_ADMIN` on project `P1`

From the UI, there should probably not be a screen for the roles, they should be added from the user side and/or the project side.


UI on the project side:

Project `P1`

Team:
`A1`  - `PROJECT_ADMIN` and `ANALYST`
`A2` - `RESEARCHER`

UI on the user side

Account `A1`

Roles:
Project `P1`: `PROJECT_ADMIN` and `ANALYST`
Project `P2`: `ANALYST`

**Comment by dennyverbeeck** [comment]
Indeed, basically a Role is a triple Userid, Projectid, Authorityid. See also the datamodel [here](https://docs.google.com/drawings/d/1CqSPZcImbxYfxSpWWkxzxpxbPFrWzM6CmB_pZ_F6F68/edit). JHipster creates screens for all entities, so i guess that's why it's there for now. I agree that in the future the creation of role should happen transparently.

**Comment by blootsvoets** [comment]
Specifically, in <https://github.com/RADAR-CNS/ManagementPortal/blob/dev/src/main/java/org/radarcns/management/domain/Role.java>, I would expect a `User user` field instead of a `Set<User> users` field, as indeed defined in the datamodel.

**Comment by nivemaham** [comment]
I have few corrections to add here. 
@dennyverbeeck A Role is **not** a triple of  Userid, Projectid, Authorityid. According to current model, a Role is a authority and a projectId. A role could be held by multiple users.  (e.g.) I want multiple patients to be the `ROLE_PARTICIPANT` of `Project-A` and a user can have multple Roles. Thus it is a ManyToMany relationship. So that we cannot have `User user` in the  [ https://github.com/RADAR-CNS/ManagementPortal/blob/dev/src/main/java/org/radarcns/management/domain/Role.java](https://github.com/RADAR-CNS/ManagementPortal/blob/dev/src/main/java/org/radarcns/management/domain/Role.java) . There is indeed a`user_role` table which represents which roles are held by which user.
Plus, Role is a custom implementation based on our schema and not a part of Jhipster model neither was not added by Jhipster. I implemented it assuming it should/could be necessary for the work-flow.
However, if it complicates the work-flow we can remove the user-interfaces and handle role creation and assignment dynamically by system assuming we allow a user to be assigned to a single project.

**Comment by dennyverbeeck** [comment]
Aha ok, i based myself off the drawing, not the code. In the drawing, the `Role` table is basically a join table from `User` to `Authority`, but every relation from `User` to `Authority` has an extra property: namely the project for which the relation holds. That was how i understood this relationship initially.

A `Set<User>` field in `Role` will create an additional join table between `User` and `Role`, with an entry for each `Role` that each `User` has. Just having a `User` field would create an entry for each user and each role in the `Roles` table, making this a one-to-may relationship from `User` to `Role`, so `User`s own their `Role`s in this model. Functionally, both solutions are valid and accomplish exactly the same thing, but having a single `User` field would eliminate an extra join operation.

So basically i was looking at `Role`s as being properties of a `User`, rather than entities on their own. I'm fine with the current solution though, i don't think the extra join will cost us anything performance-wise in the real world. It could also help having `Role` normalized out if at some point we need to add additional properties to `Role`s.

**Comment by nivemaham** [comment]
True. The schema drawn here doesn't specify all the relationship tables we have explicitly. That could be a point of confusion. We could remove the Role layer if necessary. Currently, i can see it doesn't add a lot ( at least the UI). I do see we have different level of flexibility of our model. The Role as  a combination of Project and Authority would be more useful if we allow a User to be assigned to multiple projects as you mentioned. But if we allow that, we need to change the authorization layer entirely to allow specific choice of Project->Role for a token/session/login. This could be tricky

**Comment by nivemaham** [comment]
@dennyverbeeck You made me rethink how we could do this. I think the role as the Triplets may allow us to have this multiple projects, roles implementation easier. I will revisit the schema and probably make and issue to implement this.

**Comment by nivemaham** [comment]
A role is implemented as a combination project-id and authority. Currently, we have disabled deleting a role. So closing issue as what we have discussed here has been implemented
